piecrust/src/store.rs

```diff
@@ -417,179 +527,101 @@ fn write_commit<P: AsRef<Path>>(
-                    base_bytecode_path.with_extension(OBJECTCODE_EXTENSION);
-                let base_metadata_path =
-                    base_bytecode_path.with_extension(METADATA_EXTENSION);
-
-                let base_memory_dir = base.memory_dir.join(&contract_hex);
-
-                fs::hard_link(base_bytecode_path, bytecode_path)?;
-                fs::hard_link(base_module_path, module_path)?;
-                fs::hard_link(base_metadata_path, metadata_path)?;
-
-                for page_index in &elem.page_indices {
-                    // Only write the clean pages, since the dirty ones have
-                    // already been written.
-                    if !pages.contains(page_index) {
-                        let new_page_path = page_path(&memory_dir, *page_index);
-                        let base_page_path =
-                            page_path(&base_memory_dir, *page_index);
-
-                        fs::hard_link(base_page_path, new_page_path)?;
-                    }
-                }
-            }
+            // we write them to the main location
+            fs::write(bytecode_main_path, &contract_data.bytecode)?;
+            fs::write(module_main_path, &contract_data.module.serialize())?;
+            fs::write(metadata_main_path, &contract_data.metadata)?;
+            dirty = true;
         }
-    }
-
-    if let Some(base) = &directories.base {
-        for (contract, elem) in base.inner.index.iter() {
-            if !commit_contracts.contains_key(contract) {
-                let contract_hex = hex::encode(contract);
-
-                let bytecode_path =
-                    directories.bytecode_dir.join(&contract_hex);
-                let module_path =
-                    bytecode_path.with_extension(OBJECTCODE_EXTENSION);
-                let metadata_path =
-                    bytecode_path.with_extension(METADATA_EXTENSION);
-
-                let base_bytecode_path = base.bytecode_dir.join(&contract_hex);
-                let base_module_path =
-                    base_bytecode_path.with_extension(OBJECTCODE_EXTENSION);
-                let base_metadata_path =
-                    base_bytecode_path.with_extension(METADATA_EXTENSION);
-
-                let memory_dir = directories.memory_dir.join(&contract_hex);
-                let base_memory_dir = base.memory_dir.join(&contract_hex);
-
-                fs::create_dir_all(&memory_dir)?;
-
-                fs::hard_link(base_bytecode_path, bytecode_path)?;
-                fs::hard_link(base_module_path, module_path)?;
-                fs::hard_link(base_metadata_path, metadata_path)?;
-
-                for page_index in &elem.page_indices {
-                    let new_page_path = page_path(&memory_dir, *page_index);
-                    let base_page_path =
-                        page_path(&base_memory_dir, *page_index);
-
-                    fs::hard_link(base_page_path, new_page_path)?;
-                }
-            }
```

Comment:
Very nice on this removal! The power of just structuring data differently based on different assumptions is insane.